### PR TITLE
feat: burning cycles programmatically

### PR DESCRIPTION
**Pull request by ggreif**
This adds a new primitive `(prim "cyclesBurn")` and a `prim.mo` function `cyclesBurn : <system> Nat -> Nat`. It can be used to burn (some of the) canister's cycles programmatically using the system interface `ic0.cycles_burn128`.

### CODE REVIEW

**Review by luc-blaeser — APPROVED** [review]
Thank you! This looks good!

File: src/codegen/compile_enhanced.ml [hunk 1]
```diff
@@ -5385,6 +5385,18 @@ module Cycles = struct
       )
     )
 
+  let burn env =
+    Func.share_code1 Func.Always env "cycle_burn" ("cycles", I64Type) [I64Type] (fun env get_x ->
+      Stack.with_words env "dst" 4L (fun get_dst ->
```
Comment (luc-blaeser) [inline src/codegen/compile_enhanced.ml:5390]: We could use `2L` as word size is 64-bit and the result is a `u128` cycles value. 
Same for `msg_cycles_accept128` and `msg_cycles_refunded128` (my mistake).

```suggestion
      Stack.with_words env "dst" 2L (fun get_dst ->
```

File: test/run-drun/basic-cycles.mo [hunk 2]
```diff
@@ -91,6 +92,7 @@ actor a {
   public func go() : async (){
     await overflow();
     await iter();
+    assert 1000 == burn<system> 1000
```
Comment (luc-blaeser) [inline test/run-drun/basic-cycles.mo:95]: Detail: Maybe we could test burn 0 and a very large `u128` just to check the double `u64` ordering?

Comment (ggreif) [reply test/run-drun/basic-cycles.mo:95]: I can test 0, but the very large will clip to available...

Comment (luc-blaeser) [reply test/run-drun/basic-cycles.mo:95]: Yes, I thought so. Just an idea, not important.

File: src/codegen/compile_enhanced.ml (hunk 1 above)
Comment (ggreif) [reply src/codegen/compile_enhanced.ml]: Nice find, please check eed383fe7c8bb30be52f15fb70ae7c536e03e53e too @luc-blaeser

File: test/run-drun/basic-cycles.mo (hunk 2 above)
Comment (ggreif) [reply test/run-drun/basic-cycles.mo:95]: I think we are fine

**Review by crusso — COMMENTED** [review]
What about the interpreters? Did we just crash on other cycles prims or do we need to adjust?

File: src/codegen/compile_enhanced.ml [hunk 3]
```diff
@@ -5338,7 +5338,7 @@ module Cycles = struct
 
   let balance env =
     Func.share_code0 Func.Always env "cycle_balance" [I64Type] (fun env ->
-      Stack.with_words env "dst" 4L (fun get_dst ->
```
Comment (crusso) [inline src/codegen/compile_enhanced.ml:5341]: Good catch (here and below). I guess we were over stack allocating here ...

**Review by crusso — APPROVED** [review]
Looks good to me. Thanks for the quick implementation!

**Pull request merged**